cpu/stm32f4/periph/eth.c

```
@@ -0,0 +1,540 @@
+            if (data && max_len >= copy) {
+                memcpy(data, p->buffer_addr, copy);
+                max_len -= copy;
+            }
+            else if (max_len < copy) {
+                len = -1;
+            }
+            p->status = DESC_OWN;
+        }
+        _next_desc(p);
+    }
+
+    if (drop) {
+        rx_curr = p;
+    }
+
+    mutex_unlock(&receive_lock);
+
+    return len;
+}
+
+int eth_try_receive(char *data, unsigned max_len)
+{
+    return _try_receive(data, max_len, 0);
+}
+
+int eth_receive_blocking(char *data, unsigned max_len)
+{
+    return _try_receive(data, max_len, 1);
+}
+
+static void _isr(netdev2_t *netdev)
+{
+    /* if the next descriptor is owned by the CPU we can get it */
+    if (!(rx_curr->status & DESC_OWN)) {
+        netdev->event_callback(netdev, NETDEV2_EVENT_RX_COMPLETE);
+    }
+}
+
+void isr_eth(void)
+{
+    volatile unsigned tmp = ETH->DMASR;
+
+    if ((tmp & ETH_DMASR_TS)) {
+        ETH->DMASR = ETH_DMASR_TS | ETH_DMASR_NIS;
+        mutex_unlock(&_tx);
+    }
+
+    if ((tmp & ETH_DMASR_RS)) {
+        ETH->DMASR = ETH_DMASR_RS | ETH_DMASR_NIS;
+        mutex_unlock(&_rx);
+        if (_netdev2) {
+            _netdev2->event_callback(_netdev2, NETDEV2_EVENT_ISR);
+        }
+    }
+
+    /* printf("r:%x\n\n", tmp); */
+
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+}
+
+void isr_eth_wkup(void)
+{
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+}
+
+void ETH_DMA_ISR(void)
+{
+    /* clear DMA done flag */
+    int stream = eth_config.dma_stream;
+    dma_base(stream)->IFCR[dma_hl(stream)] = dma_ifc(stream);
+    mutex_unlock(&_dma_sync);
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+}
+
+static int _send(netdev2_t *netdev, const struct iovec *vector, int count)
+{
+    (void)netdev;
+
+    int ret = 0, len = 0;
+    mutex_lock(&send_lock);
+    for (int i = 0; i < count && ret == 0; i++) {
```

It returns the amount of bytes transmitted or -1 in case of error. Since netdev expect to return the amount of bytes transmitted is kind of practical to do it like this. In any case I added a condition in `_send` to make sure we don't ignore an error an pass it to netdev.